src/UserInterface.php

```diff
@@ -18,7 +18,7 @@ public function getUsername(): string;
     /**
      * Get the user role
      *
-     * @return string
+     * @return array|null
      */
-    public function getUserRole(): string;
+    public function getUserRoles(): ?array;
```

Comment:
Why should this return `null` instead of an empty array?

Reply:
A `null` value indicates that the roles are not defined. An empty array can indicate that the roles are used but the user does not have a role.

Reply:
Do you think this could make a difference for a consumer of the interface?

Reply:
@MidnightDesign not a big difference for the API but it matters from a semantic point of view, and we are also using a new feature of PHP 7.1.

Reply:
I tend to agree with @MidnightDesign here. If the value type varies, you then need to add conditionals to your consumer:

```php
$roles = $user->getUserRoles();
if (null === $roles) {
    // process differently
} else {
    foreach ($roles as $role) {
        // ...
    }
}

// or:
$roles = $user->getUserRoles() ?: [];
foreach ($roles as $role) {
    // ...
}
```

While the latter is succinct, it requires having the extra knowledge that a null value is possible, and an extra operation to ensure we have a list of roles once called. Compare this to _always_ expecting an array:

```php
foreach ($user->getUserRoles() as $role) {
     // ...
}
```

The above has the benefit of not requiring an intermediary variable prior to iteration.  Additionally, if an empty array is returned, no iteration happens at all.

This was a design decision we approached with PSR-7 as well. I'd originally advocated for null returns for empty lists, but a number of folks showed me that consistency of return type has a ton of value.